README.md
# sock

Small script to simplify network communication.

Something like [telnetlib](http://docs.python.org/library/telnetlib.html), but for clean TCP/UDP (no command sequences, \r\n newlines, etc.)

As an alternative, consider [pwntools](http://pwntools.com/) which contains unified interface (tubes) for communications with sockets, processes, etc. See [tubes API](http://pwntools.com/).

Current development only supports python 3. Old python 2 version is available at the py2 branch.

## Installation

```bash
$ pip3 install sock
```

For development or building from this repository, [poetry](https://python-poetry.org/) is needed.

## Usage

### TCP Client

```python
from sock import *

f = Sock("some.cool.servi.ce:3123", timeout=10)
# or IPv6
f = Sock6("::1 3123", timeout=3)
# or already existing socket
f = Sock.from_socket(some_socket)  # or toSock(some_socket)
# or UDP/IPv6
f = SockU6("::1 3123", timeout=3)

# wait for prompt (skip banner for example)
# the prompt itself will be skipped (and returned) too
f.read_until("> ", timeout=3)  # read_until_re also exists

f.send("flip coin\n")

# skip until regexp
result1 = f.skip_until_re(r"You've got (heads|tails)")  # skip_until(str) also exists

# read until also consumes matched part
f.read_until_re(r"You've g[oe]t ")  # read_until(str) also exists

# read specific number of bytes
result2 = f.read_nbytes(5)

assert result1 == result2

# alias for f.send(s + "\n")
f.send_line("random please")

# read one packet and flush buffers
print(f.read_one())

# non-blocking read (flush buffers)
print(f.read_one(0))

# read until disconnect
print(f.read_all())
```


About
---------------------

Author: hellman

License: [MIT License](http://opensource.org/licenses/MIT)


CHANGELOG.md
## 0.5.0 (Sep 27, 2020)

  - swtich to python 3
  - ssl wrapper

## 0.4.0 (Mar 17, 2015)

  - sock.interact() - wrapper for telnetlib interact
  - Sock.from_socket() - alias for toSock
  - read/skip \_until\_re return string instead of match if number of groups <= 1 (similar to re.findall)

## 0.3.0 (Dec 30, 2013)

  - read/skip \_until\_re now return regexp match (usefull for quick parsing)
  - added send\_line alias

## 0.2.0 (Dec 27, 2013)

  - restored first address format: Sock("ip", port)
  - added small udp support
  - removed server classes


test_sock.py
# -*- coding: utf-8 -*-

import unittest

from itertools import product

from sock import parse_addr, Sock, Sock6, SockU, SockU6


class TestParseAddr(unittest.TestCase):
    HOSTS = (
        """
        127.0.0.1  8.8.8.8  255.255.255.255  1.1.1.1
        fe80::0  FE80:0000:0000:0000:0202:B3FF:FE1E:8329  FE80::0202:B3FF:FE1E:8329  [2001:db8:0:1]
        example.com  x.x  localhost  some.bigdomain  somehost x.y.z.w.a_1._b.z
        """
    ).strip().split()

    PORTS = (
        0, 1, 2, 3, 21, 80, 443, 1024, 1337, 3123, 31337, 5000, 65534, 65535
    )

    def test_good(self):
        for host, port in product(self.HOSTS, self.PORTS):
            good = (host.strip("[]"), port)

            self.assertEqual( parse_addr(host, port), good)
            self.assertEqual( parse_addr((host, port)), good)
            self.assertEqual( parse_addr("%s:%s" % (host, port)), good)
            self.assertEqual( parse_addr("%s: %s" % (host, port)), good)
            self.assertEqual( parse_addr("%s|%s" % (host, port)), good)
            self.assertEqual( parse_addr("%s/%s" % (host, port)), good)
            self.assertEqual( parse_addr("%s %s" % (host, port)), good)
            self.assertEqual( parse_addr("%s:| /%s" % (host, port)), good)
            self.assertEqual( parse_addr("%s;%s" % (host, port)), good)

    def test_bad(self):
        for host, port in product(self.HOSTS, self.PORTS):
            self.assertRaises(TypeError, parse_addr, (host, port), port)
            self.assertRaises(TypeError, parse_addr, (host,), port)


class TestConnects(unittest.TestCase):
    HTTP_HOSTS = "google.com".strip().split()
    DNS_HOSTS = "8.8.8.8".strip().split()

    def test_http(self):
        for host in self.HTTP_HOSTS:
            f = Sock(host, 80)
            f.send("GET / HTTP/1.1\r\n\r\n")
            line = f.read_line()
            self.assertTrue(line.startswith("HTTP/1.1 "))
            f.close()

    def test_dns(self):
        query = "241a010000010000000000000377777706676f6f676c6503636f6d0000010001".decode("hex")
        marker = "\x03www\x06google\x03com"
        for host in self.DNS_HOSTS:
            f = SockU(host, 53)
            f.send(query)
            line = f.read_one()
            self.assertIn(marker, line)
            f.close()


if __name__ == "__main__":
    unittest.main()


sock.py
import re
import ssl
import sys
import socket
import telnetlib

from time import time
from socket import timeout as Timeout, error as SocketError

def Bytes(s):
    if isinstance(s, str): return s.encode("ascii")
    assert isinstance(s, bytes)
    return s

def Str(s):
    if isinstance(s, bytes): return s.decode("ascii")
    assert isinstance(s, str)
    return s

__all__ = "Sock Sock6 toSock SockU SockU6 toSockU Timeout SocketError".split()

DEFAULT_TIMEOUT = 30
PORT_REGEXP = re.compile(r"(:| |;|/|\|)+(?P<port>\d+)$")

'''
TODO:
- tests
- check Sock6/toSock6
- read_until(_re) accept list also
- think about losing socket data on EOFError
- quick fail mode? if exploit fails, read_until("$ ") will wait the whole timeout;
'''


def parse_addr(addr, port=None):
    """
    If port is embedded in addr string, it should be delimited with one or more of:
        " :;|/"

    Examples:
        parse_addr("localhost", 3123)
        parse_addr(("localhost", 3123))
        parse_addr("127.0.0.1:3123")
        parse_addr("127.0.0.1: 3123")
        parse_addr("127.0.0.1|3123")
        parse_addr("127.0.0.1/3123")
        parse_addr("127.0.0.1 3123")
        parse_addr("example.com:3123")
        parse_addr("example.com:| /3123")
    """
    _host = None
    _port = None

    if port is not None:
        _host = addr
        _port = int(port)

    elif isinstance(addr, tuple):
        _host = addr[0]
        _port = int(addr[1])

    elif isinstance(addr, str):
        match = PORT_REGEXP.search(addr.strip())
        if match:
            _host = addr[:match.start()]
            _port = int(match.group("port"))

    if _host is None or _port is None:
        raise TypeError("Can't understand address: addr=%s, port=%s" % (addr, port))

    if not isinstance(_host, str):
        raise TypeError("Host should be string")

    # strip IPv6 brackets
    _host = _host.strip("[]")
    return _host, _port


class AbstractSock(object):
    """
    SomeSock("127.0.0.1", 3123, timeout=15)
    - timeout should be given using explicit keyword
    """

    SOCKET_FAMILY = NotImplemented
    SOCKET_TYPE = NotImplemented
    RECV_SIZE = 4096

    def __init__(self, *addr, **timeout_dict):
        self.addr = parse_addr(*addr)

        # python2 does not allow (*args, timeout=None) :(
        self.timeout = float(timeout_dict.pop("timeout", DEFAULT_TIMEOUT))
        if timeout_dict:
            raise TypeError("Only timeout should be given through keyword args")

        self.buf = b""
        self.eof = False

        self.sock = socket.socket(self.SOCKET_FAMILY, self.SOCKET_TYPE)
        self.sock.settimeout(self.timeout)
        self._connect()
        return

    @classmethod
    def from_socket(cls, sock, timeout=None):
        self = object.__new__(cls)
        self.addr = sock.getpeername()
        self.sock = sock

        assert self.SOCKET_TYPE == sock.type

        if timeout is not None:
            self.timeout = float(timeout)
            self.sock.settimeout(self.timeout)
        else:
            self.timeout = self.sock.gettimeout()

        self.buf = b""
        self.eof = False
        return self

    def _connect(self):
        return NotImplemented

    def recv(self):
        return NotImplemented

    def send(self):
        return NotImplemented

    def send_line(self, line):
        return self.send(Bytes(line) + b"\n")

    def read_line(self, timeout=None):
        return self.read_until(b"\n", timeout=timeout)

    def read_one(self, timeout=None):
        """
        Read something from socket
        timeout = -1  - wait until something new in socket
        timeout = 0   - return cached buffer + socket buffer immediately
        timeout = N   - wait N seconds until something new in socket, else raise TimeoutError
        """
        self._fill_one(timeout)
        if not self.buf and timeout != 0:
            raise EOFError("Connection closed")
        res = self.buf
        self.buf = b""
        return res

    def read_all(self, timeout=None):
        """
        Read everything from socket (the other side should close socket before timeout)
        """
        self.read_cond(lambda x: x.eof, timeout)
        res = self.buf
        self.buf = b""
        return res

    def skip_until(self, s, timeout=None):
        """
        Skip everything until first occurence of string @s, stop before occurence.
        Return nothing.
        """
        s = Bytes(s)
        self.read_cond(lambda x: s in x.buf, timeout)
        start = self.buf.find(s)
        self.buf = self.buf[start:]
        return

    def skip_until_re(self, r, flags=0, timeout=None):
        """
        Skip everything until first match of regexp @r, stop before match.
        Return match.
        """
        r = Bytes(r)
        match = self.read_cond(
            lambda x: re.search(r, x.buf, flags=flags), timeout)
        self.buf = self.buf[match.start():]
        return match if len(match.groups()) > 1 else match.group(len(match.groups()))

    def read_until(self, s, timeout=None):
        """
        Read everything until first occurence of string @s, stop after occurence.
        Return everything before @s, and @s.
        """
        s = Bytes(s)
        self.read_cond(lambda x: s in x.buf, timeout)
        end = self.buf.find(s) + len(s)
        res = self.buf[:end]
        self.buf = self.buf[end:]
        return res

    def read_until_re(self, r, flags=0, timeout=None):
        """
        Read everything until first match of regexp @r, stop after match.
        Return match.
        Note: if you need the data before match, you can make group (.*?) for that:
            r1 = r"(\d) coins"
            r2 = r"(.*?)(\d coins)"
        """
        r = Bytes(r)
        match = self.read_cond(lambda x: re.search(r, x.buf, flags=flags), timeout)
        self.buf = self.buf[match.end():]
        return match if len(match.groups()) > 1 else match.group(len(match.groups()))

    def read_nbytes(self, n, timeout=None):
        self.read_cond(lambda x: len(x.buf) >= n, timeout)
        self.buf, res = self.buf[n:], self.buf[:n]
        return res

    def read_cond(self, cond, timeout=None):
        """
        Read bytes while @cond(self) is False. Return @cond return.
        self.buf should be analyzed/cropped by caller, if needed
        (this is rather low-level function, helper)
        """
        time_start = time()
        remaining = timeout
        if timeout is None:
            timeout = self.timeout

        if self.eof and not self.buf:
            raise EOFError("Connection closed")

        res = cond(self)

        while not res:
            self._fill_one(remaining)

            res = cond(self)
            if res:
                break

            if self.eof:
                raise EOFError("Connection closed")

            if timeout == -1:
                remaining = -1
            elif timeout == 0:
                raise Timeout("read_cond timeout")
            else:
                remaining = time_start + timeout - time()
                if remaining <= 0:
                    raise Timeout("read_cond timeout")

        return res

    def _fill_one(self, timeout=None):
        """Read something from socket.
        timeout = -1  -  blocking until read
        timeout = 0   -  non-blocking
        timeout = N   -  blocking until read or timeout
        """
        if timeout is None:
            timeout = self.timeout

        if timeout == 0:
            self.sock.setblocking(False)
            try:
                self.buf += self.recv(self.RECV_SIZE)
            except SocketError:
                # WHAT?
                pass
            return

        if timeout == -1:
            self.sock.settimeout(None)  # blocking, infinity timeout
        else:
            self.sock.setblocking(True)  # it's overriden by settimeout, but for clarity
            self.sock.settimeout(timeout)

        buf = self.recv(self.RECV_SIZE)
        self.eof = (not buf)
        self.buf += buf
        return

    @property
    def socket(self):
        return self.sock

    @property
    def fileno(self):
        return self.sock.fileno()

    def write(self, s):
        return self.send(s)

    def shut_wr(self):
        self.sock.shutdown(socket.SHUT_WR)

    def shut_rd(self):
        self.sock.shutdown(socket.SHUT_RD)

    def close(self):
        return self.sock.close()

    def __del__(self):
        self.sock.close()

    def interact_telnet(self):
        sys.stdout.buffer.write(self.buf)
        self.buf = b""

        t = telnetlib.Telnet()
        t.sock = self.sock
        return t.interact()

    def interact(self):
        # copied from Telnetlib with minor fixes
        import selectors
        _TelnetSelector = selectors.SelectSelector

        sys.stdout.buffer.write(self.buf)
        self.buf = b""

        with _TelnetSelector() as selector:
            selector.register(self.sock, selectors.EVENT_READ)
            selector.register(sys.stdin, selectors.EVENT_READ)

            while True:
                for key, events in selector.select():
                    if key.fileobj is self.sock:
                        try:
                            text = self.read_one()
                        except EOFError:
                            print('*** Connection closed by remote host ***')
                            return
                        if text:
                            sys.stdout.buffer.write(text)
                            sys.stdout.flush()
                    elif key.fileobj is sys.stdin:
                        line = sys.stdin.readline()
                        if not line:
                            return
                        self.send(line)


class IPv4Mixin(object):
    SOCKET_FAMILY = socket.AF_INET


class IPv6Mixin(object):
    SOCKET_FAMILY = socket.AF_INET6


class TCPMixIn(object):
    SOCKET_TYPE = socket.SOCK_STREAM

    def _connect(self):
        self.sock.connect(self.addr)

    def recv(self, bufsize):
        return self.sock.recv(bufsize)

    def send(self, s):
        return self.sock.sendall(Bytes(s))

class SSLMixIn(TCPMixIn):
    def _connect(self):
        TCPMixIn._connect(self)
        self.sock = ssl.wrap_socket(self.sock)


class UDPMixIn(object):
    SOCKET_TYPE = socket.SOCK_DGRAM

    def _connect(self):
        pass  # udp doesn't need connect

    def recv(self, bufsize):
        while True:
            data, addr = self.sock.recvfrom(bufsize)
            if addr == self.addr:
                return data
            # TODO: warning about non-matching addr

    def send(self, s):
        return self.sock.sendto(s, self.addr)


class Sock(TCPMixIn, IPv4Mixin, AbstractSock):
    pass


class Sock6(TCPMixIn, IPv6Mixin, AbstractSock):
    pass


class SockU(UDPMixIn, IPv4Mixin, AbstractSock):
    pass


class SockU6(UDPMixIn, IPv6Mixin, AbstractSock):
    pass


class SSLSock(SSLMixIn, IPv4Mixin, AbstractSock):
    pass


class SSLSock6(SSLMixIn, IPv6Mixin, AbstractSock):
    pass

toSock = Sock.from_socket
toSockU = SockU.from_socket


server.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

import os
import sys
import time
import threading
from socket import *

__all__ = "tcp tcp6 udp udp6".split()

# IPv4/IPv6 --------------------

class inet:
    SOCKET_FAMILY = AF_INET
    DEFAULT_HOST = "0.0.0.0"

    def __init__(self, port, handler, host=None):
        self.port = int(port)
        self.handler = handler
        self.host = host if host else self.DEFAULT_HOST
        self.socket = socket(self.SOCKET_FAMILY, self.SOCKET_TYPE)
        self.socket.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
        self.socket.bind((self.host, self.port))

class inet6(inet):
    SOCKET_FAMILY = AF_INET6
    DEFAULT_HOST = "::"


# UDP --------------------------
# udp handler accepts (client, data, addr) arguments

class udp(inet):
    SOCKET_TYPE = SOCK_DGRAM
    def listen(self, maxthreads=10, debug=True):
        if debug: print "[i] %s Started UDP server on [%s]:%d" % (time.asctime(), self.host, self.port)
        while True:
            data, addr = self.socket.recvfrom(4096)
            if debug: print "[i]", time.asctime(), "Connect from %s" % (addr, )
            
            while threading.active_count() - 1 >= maxthreads:
                if debug: sys.stderr.write("[W] WARNING: Too many threads! %d\n" % threading.active_count())
                time.sleep(1)

            t = threading.Thread(target=self.handler, args=(self.socket, data, addr))
            t.setDaemon(True)
            t.start()
        return

class udp6(inet6, udp):
    pass


# TCP --------------------------
# tcp handler accepts (client, addr) arguments

class tcp(inet):
    SOCKET_TYPE = SOCK_STREAM
    def listen(self, maxthreads=1000, maxconn=100, debug=True):
        self.socket.listen(maxconn)
        if debug: print "[i] %s Started TCP server on [%s]%d" % (time.asctime(), self.host, self.port)
        while True:
            client, addr = self.socket.accept()
            if debug: print "[i]", time.asctime(), "Connect from", addr
            
            while threading.active_count() - 1 >= maxthreads:
                if debug: sys.stderr.write("[W] WARNING: Too many threads! %d\n" % threading.active_count())
                time.sleep(0.5)

            t = threading.Thread(target=self.handler, args=(client, addr))
            t.setDaemon(True)
            t.start()
        return

class tcp6(inet6, tcp):
    pass

sock3.py
# python3-version of Sock

import re
import ssl
import sys
import socket
import telnetlib

from time import time
from socket import timeout as Timeout, error as SocketError

def Bytes(s):
    if isinstance(s, str): return s.encode("ascii")
    assert isinstance(s, bytes)
    return s

def Str(s):
    if isinstance(s, bytes): return s.decode("ascii")
    assert isinstance(s, str)
    return s

__all__ = "Sock Sock6 toSock SockU SockU6 toSockU Timeout SocketError".split()

DEFAULT_TIMEOUT = 30
PORT_REGEXP = re.compile(r"(:| |;|/|\|)+(?P<port>\d+)$")

'''
TODO:
- tests
- check Sock6/toSock6
- read_until(_re) accept list also
- think about losing socket data on EOFError
- quick fail mode? if exploit fails, read_until("$ ") will wait the whole timeout;
- pexpect like read_until ( https://pexpect.readthedocs.org/en/latest/api/pexpect.html )
'''


def parse_addr(addr, port=None):
    """
    If port is embedded in addr string, it should be delimited with one or more of:
        " :;|/"

    Examples:
        parse_addr("localhost", 3123)
        parse_addr(("localhost", 3123))
        parse_addr("127.0.0.1:3123")
        parse_addr("127.0.0.1: 3123")
        parse_addr("127.0.0.1|3123")
        parse_addr("127.0.0.1/3123")
        parse_addr("127.0.0.1 3123")
        parse_addr("example.com:3123")
        parse_addr("example.com:| /3123")
    """
    _host = None
    _port = None

    if port is not None:
        _host = addr
        _port = int(port)

    elif isinstance(addr, tuple):
        _host = addr[0]
        _port = int(addr[1])

    elif isinstance(addr, str):
        match = PORT_REGEXP.search(addr.strip())
        if match:
            _host = addr[:match.start()]
            _port = int(match.group("port"))

    if _host is None or _port is None:
        raise TypeError("Can't understand address: addr=%s, port=%s" % (addr, port))

    if not isinstance(_host, str):
        raise TypeError("Host should be string")

    # strip IPv6 brackets
    _host = _host.strip("[]")
    return _host, _port


class AbstractSock(object):
    """
    SomeSock("127.0.0.1", 3123, timeout=15)
    - timeout should be given using explicit keyword
    """

    SOCKET_FAMILY = NotImplemented
    SOCKET_TYPE = NotImplemented
    RECV_SIZE = 4096

    def __init__(self, *addr, **timeout_dict):
        self.addr = parse_addr(*addr)

        # python2 does not allow (*args, timeout=None) :(
        self.timeout = float(timeout_dict.pop("timeout", DEFAULT_TIMEOUT))
        if timeout_dict:
            raise TypeError("Only timeout should be given through keyword args")

        self.buf = b""
        self.eof = False

        self.sock = socket.socket(self.SOCKET_FAMILY, self.SOCKET_TYPE)
        self.sock.settimeout(self.timeout)
        self._connect()
        return

    @classmethod
    def from_socket(cls, sock, timeout=None):
        self = object.__new__(cls)
        self.addr = sock.getpeername()
        self.sock = sock

        assert self.SOCKET_TYPE == sock.type

        if timeout is not None:
            self.timeout = float(timeout)
            self.sock.settimeout(self.timeout)
        else:
            self.timeout = self.sock.gettimeout()

        self.buf = b""
        self.eof = False
        return self

    def _connect(self):
        return NotImplemented

    def recv(self):
        return NotImplemented

    def send(self):
        return NotImplemented

    def send_line(self, line):
        return self.send(Bytes(line) + b"\n")

    def read_line(self, timeout=None):
        return self.read_until(b"\n", timeout=timeout)

    def read_one(self, timeout=None):
        """
        Read something from socket
        timeout = -1  - wait until something new in socket
        timeout = 0   - return cached buffer + socket buffer immediately
        timeout = N   - wait N seconds until something new in socket, else raise TimeoutError
        """
        self._fill_one(timeout)
        if not self.buf and timeout != 0:
            raise EOFError("Connection closed")
        res = self.buf
        self.buf = b""
        return res

    def read_all(self, timeout=None):
        """
        Read everything from socket (the other side should close socket before timeout)
        """
        self.read_cond(lambda x: x.eof, timeout)
        res = self.buf
        self.buf = b""
        return res

    def skip_until(self, s, timeout=None):
        """
        Skip everything until first occurence of string @s, stop before occurence.
        Return nothing.
        """
        s = Bytes(s)
        self.read_cond(lambda x: s in x.buf, timeout)
        start = self.buf.find(s)
        self.buf = self.buf[start:]
        return

    def skip_until_re(self, r, flags=0, timeout=None):
        """
        Skip everything until first match of regexp @r, stop before match.
        Return match.
        """
        r = Bytes(r)
        match = self.read_cond(
            lambda x: re.search(r, x.buf, flags=flags), timeout)
        self.buf = self.buf[match.start():]
        return match if len(match.groups()) > 1 else match.group(len(match.groups()))

    def read_until(self, s, timeout=None):
        """
        Read everything until first occurence of string @s, stop after occurence.
        Return everything before @s, and @s.
        """
        s = Bytes(s)
        self.read_cond(lambda x: s in x.buf, timeout)
        end = self.buf.find(s) + len(s)
        res = self.buf[:end]
        self.buf = self.buf[end:]
        return res

    def read_until_re(self, r, flags=0, timeout=None):
        """
        Read everything until first match of regexp @r, stop after match.
        Return match.
        Note: if you need the data before match, you can make group (.*?) for that:
            r1 = r"(\d) coins"
            r2 = r"(.*?)(\d coins)"
        """
        r = Bytes(r)
        match = self.read_cond(lambda x: re.search(r, x.buf, flags=flags), timeout)
        self.buf = self.buf[match.end():]
        return match if len(match.groups()) > 1 else match.group(len(match.groups()))

    def read_nbytes(self, n, timeout=None):
        self.read_cond(lambda x: len(x.buf) >= n, timeout)
        self.buf, res = self.buf[n:], self.buf[:n]
        return res

    def read_cond(self, cond, timeout=None):
        """
        Read bytes while @cond(self) is False. Return @cond return.
        self.buf should be analyzed/cropped by caller, if needed
        (this is rather low-level function, helper)
        """
        time_start = time()
        remaining = timeout
        if timeout is None:
            timeout = self.timeout

        if self.eof and not self.buf:
            raise EOFError("Connection closed")

        res = cond(self)

        while not res:
            self._fill_one(remaining)

            res = cond(self)
            if res:
                break

            if self.eof:
                raise EOFError("Connection closed")

            if timeout == -1:
                remaining = -1
            elif timeout == 0:
                raise Timeout("read_cond timeout")
            else:
                remaining = time_start + timeout - time()
                if remaining <= 0:
                    raise Timeout("read_cond timeout")

        return res

    def _fill_one(self, timeout=None):
        """Read something from socket.
        timeout = -1  -  blocking until read
        timeout = 0   -  non-blocking
        timeout = N   -  blocking until read or timeout
        """
        if timeout is None:
            timeout = self.timeout

        if timeout == 0:
            self.sock.setblocking(False)
            try:
                self.buf += self.recv(self.RECV_SIZE)
            except SocketError:
                # WHAT?
                pass
            return

        if timeout == -1:
            self.sock.settimeout(None)  # blocking, infinity timeout
        else:
            self.sock.setblocking(True)  # it's overriden by settimeout, but for clarity
            self.sock.settimeout(timeout)

        buf = self.recv(self.RECV_SIZE)
        self.eof = (not buf)
        self.buf += buf
        return

    @property
    def socket(self):
        return self.sock

    @property
    def fileno(self):
        return self.sock.fileno()

    def write(self, s):
        return self.send(s)

    def shut_wr(self):
        self.sock.shutdown(socket.SHUT_WR)

    def shut_rd(self):
        self.sock.shutdown(socket.SHUT_RD)

    def close(self):
        return self.sock.close()

    def __del__(self):
        self.sock.close()

    def interact_telnet(self):
        t = telnetlib.Telnet()
        sys.stdout.buffer.write(self.buf)
        t.sock = self.sock
        return t.interact()

    def interact(self):
        # copied from Telnetlib with minor fixes
        import selectors
        _TelnetSelector = selectors.SelectSelector

        sys.stdout.buffer.write(self.buf)
        with _TelnetSelector() as selector:
            selector.register(self.sock, selectors.EVENT_READ)
            selector.register(sys.stdin, selectors.EVENT_READ)

            while True:
                for key, events in selector.select():
                    if key.fileobj is self.sock:
                        try:
                            text = self.read_one()
                        except EOFError:
                            print('*** Connection closed by remote host ***')
                            return
                        if text:
                            sys.stdout.buffer.write(text)
                            sys.stdout.flush()
                    elif key.fileobj is sys.stdin:
                        line = sys.stdin.readline().encode('ascii')
                        if not line:
                            return
                        self.send(line)


class IPv4Mixin(object):
    SOCKET_FAMILY = socket.AF_INET


class IPv6Mixin(object):
    SOCKET_FAMILY = socket.AF_INET6


class TCPMixIn(object):
    SOCKET_TYPE = socket.SOCK_STREAM

    def _connect(self):
        self.sock.connect(self.addr)

    def recv(self, bufsize):
        return self.sock.recv(bufsize)

    def send(self, s):
        return self.sock.sendall(s)


class SSLMixIn(TCPMixIn):
    def _connect(self):
        TCPMixIn._connect(self)
        self.sock = ssl.wrap_socket(self.sock)


class UDPMixIn(object):
    SOCKET_TYPE = socket.SOCK_DGRAM

    def _connect(self):
        pass  # udp doesn't need connect

    def recv(self, bufsize):
        while True:
            data, addr = self.sock.recvfrom(bufsize)
            if addr == self.addr:
                return data
            # TODO: warning about non-matching addr

    def send(self, s):
        return self.sock.sendto(s, self.addr)


class Sock(TCPMixIn, IPv4Mixin, AbstractSock):
    pass


class Sock6(TCPMixIn, IPv6Mixin, AbstractSock):
    pass


class SockU(UDPMixIn, IPv4Mixin, AbstractSock):
    pass


class SockU6(UDPMixIn, IPv6Mixin, AbstractSock):
    pass


class SSLSock(SSLMixIn, IPv4Mixin, AbstractSock):
    pass


class SSLSock6(SSLMixIn, IPv6Mixin, AbstractSock):
    pass

toSock = Sock.from_socket
toSockU = SockU.from_socket
